projects/graph/graph.py

```diff
@@ -7,53 +7,157 @@ class Graph:
         beginning from starting_vertex.
         """
-        pass  # TODO
-    def dft_recursive(self, starting_vertex):
+        # Create an empty set to store visited nodes
+        visited = set()
+        # Create an empty Stack and push the starting vertex
+        s = Stack()
+        s.push(starting_vertex)
+        # While the stack is not empty...
+        while s.size() > 0:
+            # Pop the first vertex
+            v = s.pop()
+            # If that vertex has not been visited...
+            if v not in visited:
+                # Mark it as visited
+                visited.add(v)
+                print(v)
+                # Then add all of its neighbors to the top of the stack
+                for neighbor in self.vertices[v]:
+                    s.push(neighbor)
+
+
+    def dft_recursive(self, starting_vertex, visited=None):
         """
         Print each vertex in depth-first order
         beginning from starting_vertex.
         This should be done using recursion.
-        """
-        pass  # TODO
+        """   
+        if visited is None:
+            visited = set()
+        # if the node hasnt been visited mark the node as visited
+        if starting_vertex not in visited:
+            print(starting_vertex)
+            visited.add(starting_vertex)
+        # call dft_recursive on each child
+            for neighbor in self.vertices[starting_vertex]:
+                self.dft_recursive(neighbor, visited)
+
     def bfs(self, starting_vertex, destination_vertex):
         """
         Return a list containing the shortest path from
         starting_vertex to destination_vertex in
         breath-first order.
+        
+        # Create an empty set to store visited nodes
+        # Create an empty Queue and enqueue A PATH TO the starting vertex
+        
+        # While the queue is not empty...
+            # Dequeue the first PATH
+            # GRAB THE VERTEX FROM THE END OF THE PATH
+            # IF VERTEX = TARGET, RETURN PATH
+            # If that vertex has not been visited...
+                # Mark it as visited
+                # Then add A PATH TO all of its neighbors to the back of the queue
+                    # Copy the path
+                    # Append neighbor to the back of the copy
+                    # Enqueue copy
         """
-        pass  # TODO
+        visited = set()
+        q = Queue()
+        q.enqueue([starting_vertex])
+
+        while q.size() > 0:
+            path = q.dequeue()
+            
+            v = path[-1]
+
+            if v == destination_vertex:
+                return path
+            
+            if v not in visited:
+                visited.add(v)
+
+                for neighbor in self.vertices[v]:
+                    path_copy = list(path)
+                
+                    path_copy.append(neighbor)
+
+                    q.enqueue(path_copy)
+            
     def dfs(self, starting_vertex, destination_vertex):
         """
```

Comment:
nice work with the BFS / DFS on the Graph class. Isnt it crazy how similar the code is for these 2 processes? In the future we could totally implement a function that handles both cases based on a passed in parameter if we wanted!